experimental/formatted_string/src/formatted_string.rs

```diff
@@ -59,63 +75,75 @@ fn raise_annotation<F: Copy, const L: usize, const L1: usize>(
 impl<F: Copy, const L: usize> LayeredFormattedStringBuilder<F, L> {
     pub fn new() -> Self {
         Self {
-            chars: String::with_capacity(40),
+            bytes: Vec::with_capacity(40),
             annotations: Vec::with_capacity(40),
         }
     }
 
-    pub fn append_fsb<const L1: usize>(
+    pub fn append_layered<const L1: usize>(
         &mut self,
-        string: LayeredFormattedStringBuilder<F, L1>,
+        string: &LayeredFormattedString<F, L1>,
         field: F,
     ) -> &mut Self {
         assert_eq!(L - 1, L1);
         // len() is always a char boundary
-        self.insert_fsb_internal(self.chars.len(), string, field)
+        self.insert_layered_internal(self.bytes.len(), string, field)
     }
 
-    pub fn prepend_fsb<const L1: usize>(
+    pub fn prepend_layered<const L1: usize>(
         &mut self,
-        string: LayeredFormattedStringBuilder<F, L1>,
+        string: &LayeredFormattedString<F, L1>,
         field: F,
     ) -> &mut Self {
         assert_eq!(L - 1, L1);
         // 0 is always a char boundary
-        self.insert_fsb_internal(0, string, field)
+        self.insert_layered_internal(0, string, field)
     }
 
-    pub fn insert_fsb<const L1: usize>(
+    pub fn insert_layered<const L1: usize>(
         &mut self,
         pos: usize,
-        string: LayeredFormattedStringBuilder<F, L1>,
+        string: &LayeredFormattedString<F, L1>,
         field: F,
     ) -> Result<&mut Self, FormattedStringBuilderError> {
         assert_eq!(L - 1, L1);
-        if !self.chars.is_char_boundary(pos) {
-            Err(FormattedStringBuilderError::PositionNotCharBoundary)
+        if pos > self.bytes.len() {
+            Err(FormattedStringBuilderError::IndexOutOfBounds(pos))
+        } else if !unsafe { from_utf8_unchecked(&self.bytes) }.is_char_boundary(pos) {
```

Comment:
nit: we should have a safety comment explaining why this is okay at each invocation (even if there's a blanket one above)